aten/src/ATen/native/sparse/SparseBlasImpl.cpp

```diff
@@ -2,11 +2,193 @@
+      tiled_sizes.push_back(blocksize[1]);
+      return t.reshape(tiled_sizes).transpose(-3, -2);
+    }
+  };
+
+  // Note that sparse values are (..., b0, b1). This means that
+  // the strided input has to be "tilable" to (..., b1, x) with
+  // any x >= 1 such that all the shapes are (block) matrix product
+  // compatible. The matrix product will then have shape (..., b0, x).
+  // This in turn means the the result has to be "tilable" to
+  // (..., b0, x).
+  //
+  // These observations imply the following restrictions:
+  // 1. strided.size(-2) has to be divisible by b1.
+  // 2. result.size(-2) has to be divisible by b0.
+  // 3. both strided.size(-1) and result.size(-1)
+  //    have to be divisible by x.
+  //
+  // Restrictions 1 and 2 are trivially satisfied.
+  // Regarding restriction 3:
+  // it would make sense to take the largest possible x for better
+  // performance since it is very likely that the last dimension
+  // is contiguous. As such, this value is exactly
+  // x = strided.size(-1), since strided.size(-1) == result.size(-1)
+
+  // See the comments above. This is our x.
+  const auto outer_blocksize = n;
+
+  Blocksize strided_blocksize = {blocksize[1], outer_blocksize};
+  const auto strided_tiled = tile_tensor(strided, strided_blocksize);
+
+  // Left argument is (..., b0, b1) and right is (..., b1, x).
+  // This naturally implies the result should be "tilable" as
+  // (..., b0, x).
+  Blocksize result_blocksize = {blocksize[0], outer_blocksize};
+  auto result_tiled = tile_tensor(result, result_blocksize);
+
+  if (compressed_layout == kSparseCsr) {
+    values.unsqueeze_(-1).unsqueeze_(-1);
+  }
+
+  Tensor compressed_indices, plain_indices;
+  std::tie(compressed_indices, plain_indices) = at::sparse_csr::getCompressedPlainIndices(compressed);
+
+  // Select block rows of the strided input that intersect with the block colums of the sparse input.
+  auto strided_tiled_selected_rows = strided_tiled.index_select(-4, plain_indices);
+
+  // Now that we know which block rows intersect with which block columns,
+  // we can perform matrix products between pairs of blocks.
+  const auto pairwise_block_mm = values.unsqueeze(-3).matmul(strided_tiled_selected_rows);
+
+  // Having pairwise block matrix products stored in pairwise_block_mm,
+  // it is sufficient to sum all the block products that share the same row
+  // encoded in the sparse index. Since the reduction step is done via
+  // advanced indexing methods, the compressed index ought to get converted
+  // to the COO format.
+  const auto compressed_indices_coo = at::_convert_indices_from_csr_to_coo(
+      compressed_indices,
+      plain_indices,
+      compressed_indices.scalar_type() == kInt).select(0, 0);
+
+  // Reduction step.
+  // Zero out and sum over the blocks that share the same row indices.
+  result_tiled.zero_();
+  result_tiled.index_add_(
+      /*dim=*/-4,
+      /*index=*/compressed_indices_coo,
+      /*source=*/pairwise_block_mm);
```

Comment:
Hmm, this is the part that hinders precision for half dtypes. Alternatively, I could do `index_select` into a promoted dtype with a follow-up down-casting to `result.dtype`. What are your thoughts on that one, @cpuhrsch , @amjames , @pearu ?

Reply:
The CUDA documentation states that for float16/bfloat16 the compute type is always the corresponding 32bit float type. The compute type is the data type used to compute the size of the intermediate workspace buffer. 

> NOTE: `CUDA_R_16F`, `CUDA_R_16BF`, `CUDA_C_16F`, and `CUDA_C_16BF` data types always imply mixed-precision computation. 


I think that having an intermediate result at a higher precision and casting to the desired output data type is completely reasonable.

Reply:
OK. Done! In the tests was able to reduce tolerance from 1e-1 to 1e-3. Given that data is randomly sampled from [0, 1), this tolerance now shows that we are doing something reasonable :)